parser/impl/parser.cpp

```diff
@@ -7,51 +7,289 @@ namespace parser
 
     // ---- @kureduro
 
+    static lexer::TokenType primitives[] = {
+        lexer::TokenType::IntegerType,
+        lexer::TokenType::RealType,
+    }; 
 
     sPtr<ast::Program> Parser::parseProgram() {
-        return std::make_shared<ast::Program>();
+        ast::Program programNode;
+        lexer::Token currentToken;
+        while ((currentToken = m_lexer.Peek()).type != lexer::TokenType::Eof) {
+            switch(currentToken.type) {
+            case lexer::TokenType::Routine:
+                programNode.routines.push_back(parseRoutine());
+                continue;
+            case lexer::TokenType::Var:
+                programNode.variables.push_back(parseVariable());
+                continue;
+            case lexer::TokenType::Type:
+                programNode.types.push_back(parseType());
+                continue;
+            case lexer::TokenType::NewLine:
+                m_lexer.Next();
+                continue;
+            default:
+                m_errors.push_back(Error{
+                    .pos = currentToken.pos,
+                    .message = "Unexpected token",
+                });
+                while(m_lexer.Peek().type != lexer::TokenType::NewLine) m_lexer.Next();
+            }
+        }
+        return std::make_shared<ast::Program>(programNode);
+    }
+
+    sPtr<ast::Routine> Parser::parseRoutine() {
+        ast::Routine routineNode;
+        lexer::Token currentToken = m_lexer.Next();
+        if (currentToken.type != lexer::TokenType::Routine) {
+            m_errors.push_back(Error{
+                .pos = currentToken.pos,
+                .message = "Expected 'routine' keyword but did not find one",
+            });
+            return nullptr;
+        }
+        currentToken = skipWhile(isNewLine);
+
+        if (currentToken.type != lexer::TokenType::Identifier) {
+            m_errors.push_back(Error{
+                .pos = currentToken.pos,
+                .message = "Expected an identifier but did not find one",
+            });
+            // skip till "end" keyword
+            while(m_lexer.Next().type != lexer::TokenType::End);
+            return nullptr;
+        }
+        routineNode.name = currentToken;
+
+        currentToken = skipWhile(isNewLine);
+
+        if (currentToken.type != lexer::TokenType::OpenParen) {
+            m_errors.push_back(Error{
+                .pos = currentToken.pos,
+                .message = "Expected to find '('",
+            });
+            // skip till "end" keyword
+            while(m_lexer.Next().type != lexer::TokenType::End);
+            return nullptr;
+        }
+
+        while(currentToken.type != lexer::TokenType::CloseParen) {
+            routineNode.parameters.push_back(parseParameter());
+            currentToken = m_lexer.Next();
+            if (currentToken.type == lexer::TokenType::Comma) {
```

Comment:
As far, as I understand, it will work even if we have a sequence of parameters, which aren't separated by comma, but separated by space.

Reply:
Right. Fixed